File: unity/Tests/StreamingTests.cs
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;
using NUnit.Framework;

namespace Ensoul.Tests
{
    [TestFixture]
    public class StreamingTests
    {
        private static SseStream MakeStream(string sseText)
        {
            var bytes = Encoding.UTF8.GetBytes(sseText);
            var content = new ByteArrayContent(bytes);
            content.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue("text/event-stream");
            var response = new HttpResponseMessage(HttpStatusCode.OK) { Content = content };
            return new SseStream(response);
        }

        private static async Task<List<SseEvent>> CollectEvents(SseStream stream)
        {
            var events = new List<SseEvent>();
            await foreach (var evt in stream.Events())
                events.Add(evt);
            return events;
        }

        [Test]
        public async Task Events_ParsesBasicEvent()
        {
            using var stream = MakeStream("data: hello\n\n");
            var events = await CollectEvents(stream);
            Assert.That(events, Has.Count.EqualTo(1));
            Assert.That(events[0].Data, Is.EqualTo("hello"));
        }

        [Test]
        public async Task Events_ParsesMultipleEvents()
        {
            using var stream = MakeStream("data: first\n\ndata: second\n\n");
            var events = await CollectEvents(stream);
            Assert.That(events, Has.Count.EqualTo(2));
            Assert.That(events[0].Data, Is.EqualTo("first"));
            Assert.That(events[1].Data, Is.EqualTo("second"));
        }

        [Test]
        public async Task Events_HandlesMultiLineData()
        {
            // Two data lines should be joined with newline
            using var stream = MakeStream("data: line1\ndata: line2\n\n");
            var events = await CollectEvents(stream);
            Assert.That(events, Has.Count.EqualTo(1));
            Assert.That(events[0].Data, Is.EqualTo("line1\nline2"));
        }

        [Test]
        public async Task Events_IgnoresComments()
        {
            using var stream = MakeStream(": this is a comment\ndata: hello\n\n");
            var events = await CollectEvents(stream);
            Assert.That(events, Has.Count.EqualTo(1));
            Assert.That(events[0].Data, Is.EqualTo("hello"));
        }

        [Test]
        public async Task Events_ParsesEventType()
        {
            using var stream = MakeStream("event: chunk\ndata: payload\n\n");
            var events = await CollectEvents(stream);
            Assert.That(events, Has.Count.EqualTo(1));
            Assert.That(events[0].Event, Is.EqualTo("chunk"));
            Assert.That(events[0].Data, Is.EqualTo("payload"));
        }

        [Test]
        public async Task Events_ParsesId()
        {
            using var stream = MakeStream("id: 42\ndata: hello\n\n");
            var events = await CollectEvents(stream);
            Assert.That(events, Has.Count.EqualTo(1));
            Assert.That(events[0].Id, Is.EqualTo("42"));
        }

        [Test]
        public async Task Events_ParsesRetry()
        {
            using var stream = MakeStream("retry: 5000\ndata: hello\n\n");
            var events = await CollectEvents(stream);
            Assert.That(events, Has.Count.EqualTo(1));
            Assert.That(events[0].Retry, Is.EqualTo(5000));
        }

        [Test]
        public async Task Events_DefaultEventIsMessage()
        {
            using var stream = MakeStream("data: hello\n\n");
            var events = await CollectEvents(stream);
            Assert.That(events[0].Event, Is.EqualTo("message"));
        }

        [Test]
        public async Task Events_BlankLineDispatchesEvent()
        {
            // Should get 2 events due to blank line separator
            using var stream = MakeStream("data: one\n\ndata: two\n\n");
            var events = await CollectEvents(stream);
            Assert.That(events, Has.Count.EqualTo(2));
        }

        [Test]
        public async Task Events_NoTrailingBlankLine_StillDispatches()
        {
            // Without trailing \n\n, event should still emit at EOF
            using var stream = MakeStream("data: hello");
            var events = await CollectEvents(stream);
            Assert.That(events, Has.Count.EqualTo(1));
            Assert.That(events[0].Data, Is.EqualTo("hello"));
        }

        [Test]
        public async Task Events_EmptyDataLines_AreIncluded()
        {
            // "data:" with no value → empty string data line
            using var stream = MakeStream("data:\ndata: second\n\n");
            var events = await CollectEvents(stream);
            Assert.That(events, Has.Count.EqualTo(1));
            Assert.That(events[0].Data, Is.EqualTo("\nsecond"));
        }

        [Test]
        public async Task Events_FieldWithNoValue_UsesEmptyString()
        {
            // "data" with no colon → field name only, value is empty string
            using var stream = MakeStream("data\n\n");
            var events = await CollectEvents(stream);
            Assert.That(events, Has.Count.EqualTo(1));
            Assert.That(events[0].Data, Is.EqualTo(""));
        }

        [Test]
        public async Task Events_MultipleEvents_ResetState()
        {
            // Second event should not inherit event type from first
            using var stream = MakeStream("event: custom\ndata: first\n\ndata: second\n\n");
            var events = await CollectEvents(stream);
            Assert.That(events, Has.Count.EqualTo(2));
            Assert.That(events[0].Event, Is.EqualTo("custom"));
            Assert.That(events[1].Event, Is.EqualTo("message")); // reset to default
        }

        [Test]
        public async Task Events_IgnoresUnknownFields()
        {
            using var stream = MakeStream("unknown-field: value\ndata: hello\n\n");
            var events = await CollectEvents(stream);
            Assert.That(events, Has.Count.EqualTo(1));
            Assert.That(events[0].Data, Is.EqualTo("hello"));
        }

        [Test]
        public async Task ParseChatEvent_DeserializesCorrectly()
        {
            var data = @"{""chunk"": ""Hello"", ""is_final"": false, ""conversation_id"": ""c1""}";
            var sseEvent = new SseEvent { Data = data };
            var evt = SseParser.ParseChatEvent(sseEvent);
            Assert.That(evt, Is.Not.Null);
            Assert.That(evt.Chunk, Is.EqualTo("Hello"));
            Assert.That(evt.IsFinal, Is.False);
            Assert.That(evt.ConversationId, Is.EqualTo("c1"));
        }

        [Test]
        public async Task ParseChatEvent_IsFinal_True()
        {
            var data = @"{""chunk"": """", ""is_final"": true, ""conversation_id"": ""c1""}";
            var evt = SseParser.ParseChatEvent(new SseEvent { Data = data });
            Assert.That(evt.IsFinal, Is.True);
        }

        [Test]
        public async Task ParseChatEvent_WithTokenUsage()
        {
            var data = @"{""chunk"": ""done"", ""is_final"": true, ""token_usage"": {""input_tokens"": 10, ""output_tokens"": 20, ""total_tokens"": 30}}";
            var evt = SseParser.ParseChatEvent(new SseEvent { Data = data });
            Assert.That(evt.TokenUsage, Is.Not.Null);
            Assert.That(evt.TokenUsage!["input_tokens"], Is.EqualTo(10));
            Assert.That(evt.TokenUsage!["total_tokens"], Is.EqualTo(30));
        }

        [Test]
        public async Task ParseAggregateEvent_DeserializesCorrectly()
        {
            var data = @"{""tally"": {""yes"": 5}, ""is_final"": false}";
            var evt = SseParser.ParseAggregateEvent(new SseEvent { Data = data });
            Assert.That(evt, Is.Not.Null);
            Assert.That(evt.IsFinal, Is.False);
        }

        [Test]
        public async Task ParseAggregateEvent_TallyValues()
        {
            var data = @"{""tally"": {""agree"": 8, ""disagree"": 2}, ""is_final"": true}";
            var evt = SseParser.ParseAggregateEvent(new SseEvent { Data = data });
            Assert.That(evt.Tally, Is.Not.Null);
            Assert.That(evt.Tally["agree"], Is.EqualTo(8));
            Assert.That(evt.IsFinal, Is.True);
        }

        [Test]
        public void Dispose_DisposesResponse()
        {
            var response = new HttpResponseMessage(HttpStatusCode.OK)
            {
                Content = new StringContent("")
            };
            var stream = new SseStream(response);
            Assert.DoesNotThrow(() => stream.Dispose());
            // Double dispose should also not throw
            Assert.DoesNotThrow(() => stream.Dispose());
        }
    }
}

File: swift/Tests/EnsoulTests/StreamingTests.swift
/// Tests for the SSE parser, typed event parsers, and related streaming types.
///
/// These tests exercise the parsing logic directly without a real HTTP connection.
/// The SSEStream AsyncIterator's `processLine` logic is tested indirectly via
/// the helper `parseSSEText()` that feeds raw SSE text through the stream.
import XCTest
@testable import Ensoul

// MARK: - SSE text-to-events helper

/// Drive the SSE parser with a raw text string that represents one or more SSE events.
///
/// This works by constructing a URLRequest that MockURLProtocol will fulfil with
/// the raw bytes of `sseText`, then iterating the resulting SSEStream.
@available(iOS 15.0, macOS 12.0, *)
private func parseSSEText(_ sseText: String) async throws -> [SSEEvent] {
    let session = MockURLProtocol.makeSession()
    MockURLProtocol.requestHandler = { request in
        let url = request.url ?? URL(string: "https://api.ensoul-ai.com/v1/test-stream")!
        let response = HTTPURLResponse(
            url: url,
            statusCode: 200,
            httpVersion: "HTTP/1.1",
            headerFields: ["Content-Type": "text/event-stream"]
        )!
        return (response, Data(sseText.utf8))
    }

    let url = URL(string: "https://api.ensoul-ai.com/v1/test-stream")!
    var request = URLRequest(url: url)
    request.httpMethod = "GET"
    let stream = SSEStream(request: request, session: session)

    var events: [SSEEvent] = []
    for try await event in stream {
        events.append(event)
    }
    return events
}

// MARK: - StreamingTests

@available(iOS 15.0, macOS 12.0, *)
final class StreamingTests: XCTestCase {

    // MARK: - Basic SSE event parsing

    func test_sseParser_simpleEvent_parsesEventAndData() async throws {
        let sseText = "event: chunk\ndata: hello\n\n"
        let events = try await parseSSEText(sseText)

        XCTAssertEqual(events.count, 1)
        XCTAssertEqual(events[0].event, "chunk")
        XCTAssertEqual(events[0].data, "hello")
    }

    func test_sseParser_defaultEventType_isMessage() async throws {
        // When no `event:` field is present, the event type defaults to "message"
        let sseText = "data: some data\n\n"
        let events = try await parseSSEText(sseText)

        XCTAssertEqual(events.count, 1)
        XCTAssertEqual(events[0].event, "message")
        XCTAssertEqual(events[0].data, "some data")
    }

    func test_sseParser_multipleEvents_parsesAll() async throws {
        let sseText = """
        event: chunk
        data: first

        event: chunk
        data: second

        event: final
        data: done

        """
        let events = try await parseSSEText(sseText)

        XCTAssertEqual(events.count, 3)
        XCTAssertEqual(events[0].data, "first")
        XCTAssertEqual(events[1].data, "second")
        XCTAssertEqual(events[2].data, "done")
        XCTAssertEqual(events[2].event, "final")
    }

    func test_sseParser_commentLines_areIgnored() async throws {
        // Lines starting with `:` are SSE comments
        let sseText = ": this is a comment\ndata: real data\n\n"
        let events = try await parseSSEText(sseText)

        XCTAssertEqual(events.count, 1)
        XCTAssertEqual(events[0].data, "real data")
    }

    func test_sseParser_multipleCommentLines_areAllIgnored() async throws {
        let sseText = ": comment one\n: comment two\ndata: payload\n\n"
        let events = try await parseSSEText(sseText)

        XCTAssertEqual(events.count, 1)
        XCTAssertEqual(events[0].data, "payload")
    }

    func test_sseParser_multilineData_joinedWithNewline() async throws {
        // Multiple `data:` lines within one event are joined with \n
        let sseText = "data: line one\ndata: line two\ndata: line three\n\n"
        let events = try await parseSSEText(sseText)

        XCTAssertEqual(events.count, 1)
        XCTAssertEqual(events[0].data, "line one\nline two\nline three")
    }

    func test_sseParser_emptyData_doesNotDispatchEvent() async throws {
        // A blank-line with no accumulated data should not emit an event
        let sseText = "\n\n"
        let events = try await parseSSEText(sseText)

        XCTAssertEqual(events.count, 0)
    }

    func test_sseParser_idField_isPopulated() async throws {
        let sseText = "id: evt_001\ndata: payload\n\n"
        let events = try await parseSSEText(sseText)

        XCTAssertEqual(events.count, 1)
        XCTAssertEqual(events[0].id, "evt_001")
    }

    func test_sseParser_noIdField_isNil() async throws {
        let sseText = "data: no-id\n\n"
        let events = try await parseSSEText(sseText)

        XCTAssertNil(events[0].id)
    }

    func test_sseParser_retryField_isParsed() async throws {
        let sseText = "retry: 3000\ndata: with-retry\n\n"
        let events = try await parseSSEText(sseText)

        XCTAssertEqual(events[0].retry, 3000)
    }

    func test_sseParser_leadingSpaceInDataValue_isStripped() async throws {
        // SSE spec: strip one leading space after the colon
        let sseText = "data: hello world\n\n"
        let events = try await parseSSEText(sseText)

        XCTAssertEqual(events[0].data, "hello world")
    }

    // MARK: - SSEEvent type

    func test_sseEvent_init_storesAllFields() {
        let event = SSEEvent(event: "chunk", data: "payload", id: "001", retry: 1000)
        XCTAssertEqual(event.event, "chunk")
        XCTAssertEqual(event.data, "payload")
        XCTAssertEqual(event.id, "001")
        XCTAssertEqual(event.retry, 1000)
    }

    func test_sseEvent_optionalFieldsDefaultToNil() {
        let event = SSEEvent(event: "message", data: "hello")
        XCTAssertNil(event.id)
        XCTAssertNil(event.retry)
    }

    // MARK: - parseChatEvent

    func test_parseChatEvent_validPayload_returnsTypedEvent() throws {
        let payload: [String: Any] = [
            "chunk": "Hello, world!",
            "conversation_id": "conv_test_001",
            "chunk_index": 0,
            "is_final": false,
        ]
        let jsonString = String(data: MockURLProtocol.jsonData(payload), encoding: .utf8)!
        let rawEvent = SSEEvent(event: "chunk", data: jsonString)

        let chatEvent = try parseChatEvent(rawEvent)

        XCTAssertEqual(chatEvent.chunk, "Hello, world!")
        XCTAssertEqual(chatEvent.conversationId, "conv_test_001")
        XCTAssertEqual(chatEvent.chunkIndex, 0)
        XCTAssertFalse(chatEvent.isFinal)
        XCTAssertNil(chatEvent.tokenUsage)
    }

    func test_parseChatEvent_finalChunk_includesTokenUsage() throws {
        let payload: [String: Any] = [
            "chunk": "done.",
            "conversation_id": "conv_test_001",
            "chunk_index": 4,
            "is_final": true,
            "token_usage": ["input_tokens": 256, "output_tokens": 42, "total_tokens": 298],
        ]
        let jsonString = String(data: MockURLProtocol.jsonData(payload), encoding: .utf8)!
        let rawEvent = SSEEvent(event: "chunk", data: jsonString)

        let chatEvent = try parseChatEvent(rawEvent)

        XCTAssertTrue(chatEvent.isFinal)
        XCTAssertEqual(chatEvent.tokenUsage?["total_tokens"], 298)
    }

    func test_parseChatEvent_invalidJSON_throwsError() throws {
        let rawEvent = SSEEvent(event: "chunk", data: "this is not json")
        XCTAssertThrowsError(try parseChatEvent(rawEvent)) { error in
            XCTAssertTrue(error is EnsoulSDKError)
        }
    }

    func test_parseChatEvent_missingRequiredFields_throwsError() throws {
        // Missing conversation_id and chunk_index
        let payload: [String: Any] = ["chunk": "partial"]
        let jsonString = String(data: MockURLProtocol.jsonData(payload), encoding: .utf8)!
        let rawEvent = SSEEvent(event: "chunk", data: jsonString)

        XCTAssertThrowsError(try parseChatEvent(rawEvent)) { error in
            XCTAssertTrue(error is EnsoulSDKError)
        }
    }

    // MARK: - parseAggregateEvent

    func test_parseAggregateEvent_validPayload_returnsTypedEvent() throws {
        let payload: [String: Any] = [
            "tally": ["positive": 12, "negative": 5, "neutral": 3],
            "n": 20,
            "categories": [["category": "positive", "count": 12] as [String: Any]],
            "can_terminate": false,
            "is_final": false,
        ]
        let jsonString = String(data: MockURLProtocol.jsonData(payload), encoding: .utf8)!
        let rawEvent = SSEEvent(event: "progress", data: jsonString)

        let aggEvent = try parseAggregateEvent(rawEvent)

        XCTAssertEqual(aggEvent.n, 20)
        XCTAssertEqual(aggEvent.tally["positive"], 12)
        XCTAssertEqual(aggEvent.tally["negative"], 5)
        XCTAssertFalse(aggEvent.canTerminate)
        XCTAssertFalse(aggEvent.isFinal)
        XCTAssertNil(aggEvent.synthesis)
    }

    func test_parseAggregateEvent_finalEvent_includesSynthesis() throws {
        let payload: [String: Any] = [
            "tally": ["positive": 145, "negative": 72, "neutral": 83],
            "n": 300,
            "categories": [] as [[String: Any]],
            "can_terminate": true,
            "is_final": true,
            "synthesis": "The majority express positive views.",
        ]
        let jsonString = String(data: MockURLProtocol.jsonData(payload), encoding: .utf8)!
        let rawEvent = SSEEvent(event: "progress", data: jsonString)

        let aggEvent = try parseAggregateEvent(rawEvent)

        XCTAssertTrue(aggEvent.isFinal)
        XCTAssertTrue(aggEvent.canTerminate)
        XCTAssertEqual(aggEvent.synthesis, "The majority express positive views.")
    }

    func test_parseAggregateEvent_invalidJSON_throwsError() throws {
        let rawEvent = SSEEvent(event: "progress", data: "bad json {")
        XCTAssertThrowsError(try parseAggregateEvent(rawEvent)) { error in
            XCTAssertTrue(error is EnsoulSDKError)
        }
    }

    func test_parseAggregateEvent_missingRequiredFields_throwsError() throws {
        // Missing tally and n
        let payload: [String: Any] = ["is_final": false]
        let jsonString = String(data: MockURLProtocol.jsonData(payload), encoding: .utf8)!
        let rawEvent = SSEEvent(event: "progress", data: jsonString)

        XCTAssertThrowsError(try parseAggregateEvent(rawEvent)) { error in
            XCTAssertTrue(error is EnsoulSDKError)
        }
    }

    // MARK: - ChatStreamEvent

    func test_chatStreamEvent_storesAllFields() {
        let event = ChatStreamEvent(
            chunk: "Hello",
            conversationId: "conv_001",
            chunkIndex: 0,
            isFinal: true,
            tokenUsage: ["total_tokens": 10]
        )
        XCTAssertEqual(event.chunk, "Hello")
        XCTAssertEqual(event.conversationId, "conv_001")
        XCTAssertEqual(event.chunkIndex, 0)
        XCTAssertTrue(event.isFinal)
        XCTAssertEqual(event.tokenUsage?["total_tokens"], 10)
    }

    // MARK: - AggregateStreamEvent

    func test_aggregateStreamEvent_storesAllFields() {
        let event = AggregateStreamEvent(
            tally: ["yes": 50, "no": 50],
            n: 100,
            categories: [["category": "yes"]],
            canTerminate: false,
            isFinal: false,
            synthesis: nil,
            extra: ["query_id": "q1"]
        )
        XCTAssertEqual(event.n, 100)
        XCTAssertEqual(event.tally["yes"], 50)
        XCTAssertFalse(event.canTerminate)
        XCTAssertFalse(event.isFinal)
    }
}
